### DISCUSSION TITLE
Nested x-for loops

### DISCUSSION THREAD

**Discussion by lliendo** [alpinejs/alpine#3788] [1. Help] [answered]
Hi,

I've been trying to make a nested `x-for` work without any luck so far. I can't understand nor see what I'm doing wrong. Below I add two snippets. The first one is regular JS code with a JS object that prints to console the data I need to basically render in the HTML template. 

The last snippet is the actual AlpineJS code that I'm using. I've checked the data, it's there in the format that I'm expecting, yet the inner `x-for` refuses to render and no errors appear in the console. I know that this feature was discussed in [Issue 158](https://github.com/alpinejs/alpine/issues/158) and the last snippet presented in that thread is actually the same thing I'm trying to do.

```JS
let highlights = {
  "decisione commissione europea 2007/589/ec 18 luglio 2017": {
    "Decisione della Commissione Europea 2007/589/EC del 18 Luglio 2017": [
      10
    ]
  },
  "delibera 111/06": {
    "Delibera 111/06 e s.m.i": [
      23,
      74,
      75,
      77
    ],
    "Delibera 111/06 e s.m.i.": [
      23
    ],
    "deliberazione 111/06": [
      12,
      16,
      20,
      26,
      29,
      30,
      32,
      60
    ],
    "deliberazione n: 111/06": [
      13
    ]
  },
  "legge 3 agosto 2017 n. 123": {
    "Legge del 3 Agosto 2017 n. 123.": [
      7
    ]
  }
}

for (let entity of Object.keys(highlights)) {
  console.log(`Entity ${entity}`);
  
  for (let keyword of Object.keys(highlights[entity])) {
    console.log(`Keyword: ${keyword}`);
  }
}
```

and the AlpineJS snippet:

```HTML
  <div>
    <ul>
      <template x-for="entity in Object.keys(state.highlights)">
        <li x-text="entity"></li>
        <ul>
          <template x-for="keyword in Object.keys(state.highlights[entity])">
            <li x-text="keyword"></li>
          </template>
        </ul>
      </template>
    </ul>
  </div>
```

This fragment only renders `<li x-text="entity"></li>` correctly, but the nested for is as if does not exist at all. Also on [Issue 158](https://github.com/alpinejs/alpine/issues/158) thread someone recently asked for this. I'm using Alpine 3.13.0.

Thanks in advance for any help,
Lucas.

**Comment by SimoTod** [comment, marked as answer]
an x-for template can only have 1 child. Change your code to
```html
<div>
    <ul>
      <template x-for="entity in Object.keys(state.highlights)">
        <li>
          <span x-text="entity"></span>
          <ul>
            <template x-for="keyword in Object.keys(state.highlights[entity])">
              <li x-text="keyword"></li>
            </template>
          </ul>
        </li>
      </template>
    </ul>
  </div>

**Reply by lliendo** [reply]
@SimoTod I'm a bit embarrassed, because I have had this issue in the past and managed to fix it from the documentation, but I honestly completely forgot about this (I thought something else was the source of the problem). I'm curious, is there a reason *why* `<template x-for="...">` must always comply to one child? Are there plans to avoid this limitation in the future?

Once again, thanks for the time taken to answer my mistake.

**Reply by SimoTod** [reply]
No worries 
It's because Alpine doesn't use a virtual dom and when the directive runs, it just generates additional html in your page. However the directive still needs to know which items are generated so it can update them when the status changes. To do that it use an internal key child element mapping and can only work with 1 child. Having multiple items is too complex, it would need a complete reworking of the directive and it would likely have a big hit on performance

**Comment by ekwoka** [comment]
In addition to the above, you're doing extra work in your code

```html
<div>
    <ul>
      <template x-for="(values, key) in state.highlights" :key="key">
        <li>
          <span x-text="key"></span>
          <ul>
            <template x-for="(_, keyword) in values" :key="keyword">
              <li x-text="keyword"></li>
            </template>
          </ul>
        </li>
      </template>
    </ul>
  </div>
  ```
